src/utils/graph.ml

```diff
@@ -0,0 +1,301 @@
+      pp_print_vertices vs
+      pp_print_edges es
+         
+  let get_vertices: t -> vertices = fst
+
+  let get_edges: t -> edges = snd
+
+  let empty = (VSet.empty, ESet.empty)
+  (** An empty trivial graph contains no vertices and no  edges *)
+
+  let is_empty: t -> bool = fun  (vs, es) ->
+    VSet.is_empty vs 
+  (** Check if the graph is empty *)
+
+  let singleton: vertex -> t = fun v ->
+    (VSet.singleton v, ESet.empty)
+  (** returns a singleton graph *)
+
+  let is_singleton: t -> bool = fun (vs, es) ->
+    VSet.cardinal vs = 1 && ESet.cardinal es = 0
+    
+  let mk_edge: vertex -> vertex -> edge
+    = fun v1 v2 -> (v1, v2)
+  (** Make an edge from two vertices  *)
+
+  let add_vertex: t -> vertex -> t
+    = fun (vs, es) v -> (VSet.add v vs,  es) 
+  (** add avertex to a graph  *)
+
+  let add_edge: t -> edge -> t
+    = fun (vs, es) (src, tgt) ->
+    if VSet.mem src vs && VSet.mem tgt vs
+    then (vs,  ESet.add (src, tgt) es)
+    else raise IllegalGraphOperation
+  (** add an  edge to a graph  *)
+                    
+  let find_edges_of_vertex: t -> vertex -> edges
+    = fun (vs, es) v -> ESet.filter (fun e -> is_vertex_in_edge e v) es 
+                      
+  let remove_vertex: t -> vertex -> t
+    = fun (vs, es) v ->
+    (VSet.remove v vs
+    , ESet.filter (fun e -> not (is_vertex_in_edge e v)) es)           
+  (** Remove the [vertex] and its associated [edges] from the graph *)
+
+  let remove_edge:  t ->  edge ->  t
+    = fun (vs, es) e -> (vs, ESet.remove e es) 
+  (** Remove an [edge] from a graph *)                             
+
+  let non_target_vertices: t -> vertices
+    = fun (vs, es) ->
+    VSet.filter (fun v -> ESet.for_all (fun e -> not (is_vertex_target e v)) es) vs
+  (** Returns a list of all vertices that have no incoming edge  *)
+
+  let non_source_vertices: t -> vertices
+    = fun (vs, es) ->
+    VSet.filter (fun v -> ESet.for_all (fun e -> not (is_vertex_target e v)) es) vs
+  (** Returns a list of all vertices that have no outgoing edges  *)
+    
+  let connect: t -> vertex -> t = fun g v ->
+    let vs = get_vertices g in
+    (VSet.add v vs
+    , VSet.fold (fun v' es' -> ESet.add (mk_edge v v') es')
+        (get_vertices g)
+        (get_edges g))
+  (** Connect [vertex] that is already in the graph to all the other vertices in the graph *)
+
+
+  let sub_graph: t -> vertices -> t = fun g vs ->
+    ( vs
+    , ESet.filter (fun (src, tgt) -> VSet.mem src vs && VSet.mem tgt vs)
+        (get_edges g))
+  (** Gets a subgraph of given graph from a given set of vertices *)
+                                          
+  let is_point_graph: t -> bool = fun (vs, es) ->
+    ESet.is_empty es
+    
+  let union: t -> t -> t = fun (v1s, e1s) (v2s, e2s) ->
+    (VSet.union v1s v2s, ESet.union e1s e2s) 
+  (** Unions two graphs *)
+    
+  let topological_sort: t -> vertex list = fun ((vs, es) as g) ->
+    let rec r_topological_sort_helper: t -> vertex list -> vertex list
```

Comment:
The `r_` prefix in `r_topological_sort_helper` doesn’t make sense anymore. Maybe you can just call the function `topological_sort'` following extended naming conventions for auxiliary helper functions.

Reply:
i would rather prefer `topological_sort_helper` rather than `topological_sort'` 
the `'` is a very small difference and can easily be missed while reading.

Reply:
Sounds good to me too.